File: Image_Processing/Adaptive_Histogram_Equalization.py
import cv2
import numpy as np
import matplotlib.pyplot as plt

# 讀取原始影像
img = cv2.imread('..\\dataset\\data_v1\\31.jpg')

# 將影像轉換為灰度
gray_img = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)

# 使用自適應直方圖均衡化處理影像
clahe = cv2.createCLAHE(clipLimit=2.0, tileGridSize=(8, 8))
adaptive_equalized_img = clahe.apply(gray_img)

# 設置 Matplotlib 子圖
plt.figure(figsize=(10, 8))

# 原始影像和直方圖
plt.subplot(2, 2, 1)
plt.imshow(cv2.cvtColor(img, cv2.COLOR_BGR2RGB))
plt.title('Original Image')

plt.subplot(2, 2, 3)
plt.hist(gray_img.ravel(), 256, [0, 256])
plt.title('Histogram of Original Image')

# 自適應直方圖均衡化後的影像和直方圖
plt.subplot(2, 2, 2)
plt.imshow(adaptive_equalized_img, cmap='gray')
plt.title('Adaptive Histogram Equalized Image')

plt.subplot(2, 2, 4)
plt.hist(adaptive_equalized_img.ravel(), 256, [0, 256])
plt.title('Histogram of Equalized Image')

plt.show()

File: Image_Processing/Adaptive_Histogram_Equalization.m
I = imread('..\dataset\根尖合併牙周病灶\31.jpg');
I_gray = rgb2gray(I); % 將彩色影像轉換為灰度影像

I_adapt = adapthisteq(I_gray);
% I_adapt = adapthisteq(I_gray, 'ClipLimit', 0.02);
% I_adapt = adapthisteq(I_gray, 'NumTiles', [8, 8]);
% I_adapt = adapthisteq(I_gray, 'Alpha', 0.1);

subplot(2,2,1);
imshow(I);
title('Original Image');

subplot(2,2,2);
imshow(I_adapt);
title('Adaptive Histogram Equalized Image');

subplot(2,2,3);
imhist(I_gray);
title('Histogram of Original Image');

subplot(2,2,4);
imhist(I_adapt);
title('Histogram of Equalized Image');